File: dbxml/examples/java/gettingStarted/retrieveDB.java
//
// See the file LICENSE for redistribution information.
//
// Copyright (c) 2004,2009 Oracle.  All rights reserved.
//
//

package dbxml.gettingStarted;

import java.io.*;

import com.sleepycat.dbxml.*;
import com.sleepycat.db.*;

class retrieveDB
{

    private static String theContainer = "namespaceExampleData.dbxml";
    private static String theDB = "testBerkeleyDB";

    private static void usage() {
        String usageMessage = "This program retrieves data from a Berkeley DB database based on information \n";
	usageMessage += "retrieved from a document in a DB XML container. Both the DB XML container and the\n";
	usageMessage += "Berkeley DB database are stored in the same database environment. The \n";
	usageMessage += "value of the node found on the XML document is used as the Berkeley DB key for the\n";
	usageMessage += "sample data stored in the database.\n\n";

	usageMessage += "Before running this program, run exampleLoadContainer to pre-populate\n";
	usageMessage += "the DB XML container with the appropriate data. Then, use the 'buildDB' sample\n";
	usageMessage += "program to build the Berkeley DB database that this program requires.\n\n";

	usageMessage += "When you run this program, identify the directory where you told exampleLoadContainer\n";
	usageMessage += "to place the sample data:\n\n";

	usageMessage += "\t-h <dbenv directory>\n";

	usageMessage += "For example:\n";
	usageMessage += "\tjava com.sleepycat.dbxml.examples.gettingStarted.retrieveDB -h examplesEnvironment\n";

	System.out.println(usageMessage);
	System.exit( -1 );
    }

    //Utility function to clean up objects, exceptions or not,
    // containers and environments must be closed.
    private static void cleanup(myDbEnv env, XmlContainer openedContainer) {
	try {
	    if (openedContainer != null)
		openedContainer.delete();
	    if (env != null)
		env.cleanup();
	} catch (Exception e) {
	    // ignore exceptions on close
	}
    }

    public static void main(String args[])
	throws Throwable {

	File path2DbEnv = null;

	for(int i = 0; i < args.length; ++i) {
            if (args[i].startsWith("-")) {
		switch(args[i].charAt(1)) {
		case 'h':
		    path2DbEnv = new File( args[++i] );
		    break;
		default:
                    usage();
		}
            }
	}

	if (path2DbEnv == null || ! path2DbEnv.isDirectory()) {
            usage();
	}

	myDbEnv env = null;
	myDb openedDatabase = null;
	XmlContainer openedContainer = null;
	Transaction dbTxn = null;
	XmlTransaction txn = null;
	try {
	    // Open an environment
	    env = new myDbEnv(path2DbEnv);
	    XmlManager theMgr = env.getManager();

	    //Open a database in that environment
	    openedDatabase = new myDb(theDB, env.getEnvironment());

	    // open a transactional container
	    XmlContainerConfig config = new XmlContainerConfig();
	    config.setTransactional(true);
	    openedContainer =
		theMgr.openContainer(theContainer, config);

	    // Start a transaction via DB
	    dbTxn = env.getEnvironment().beginTransaction(null, null);
	    txn = theMgr.createTransaction(dbTxn);

	    //Obtain an XmlQueryContext with the result type set to
	    // values (not documents).
	    XmlQueryContext resultsContext = theMgr.createQueryContext();

	    String theQuery=
		"distinct-values(collection('namespaceExampleData.dbxml')/vendor/salesrep/name)";

	    //Get all the vendor documents out of the container
	    XmlResults results = theMgr.query(txn, theQuery,
					      resultsContext);
	    XmlValue value = results.next();
	    while (value != null) {
		    //Pull the value out of the document query result set.
		    String theSalesRepKey = value.asString();

		    DatabaseEntry theKey =
			new DatabaseEntry(theSalesRepKey.getBytes());
		    DatabaseEntry theData = new DatabaseEntry();

		    OperationStatus status = openedDatabase.getDatabase().get(txn.getTransaction(),
									      theKey, theData, null);
		    System.out.println("For key: " + theSalesRepKey +
				       ", retrieved:");
		    if (status == OperationStatus.NOTFOUND) {
			System.out.println("Key not found: run buildDB first");
		    } else
			System.out.println( new String(theData.getData(), 0,
						       theData.getSize()));
		    System.out.println("here");
		    value = results.next();
	    }
	    results.delete();
	    txn.commit();
	} catch (XmlException e) {
	    System.err.println("Xml error performing query against " + theContainer);
	    System.err.println("   Message: " + e.toString());
	    if ( txn != null ) {
		txn.abort();
	    }
	    throw e;
	}
	catch (DatabaseException de) {
	    System.err.println("Error performing query against " + theContainer);
	    System.err.println("   Message: " + de.toString());
	    if ( txn != null ) {
		txn.abort();
	    }
	    throw de;
	}
	finally {
	    if (openedDatabase != null)
		openedDatabase.cleanup();
	    cleanup(env, openedContainer);
	}
    } //End main

}

File: dbxml/examples/perl/gettingStarted/retrieveDB.pl
use strict;
use warnings;
use Sleepycat::DbXml 'simple';


sub usage()
{
    warn  <<EOM;
This program retrieves data from a Berkeley DB data based on information
retrieved from an XML document. The document is first retrieved from an
XML container. Both the database and the container are stored in the same
database environment, and all the necessary queries are performed under the
protection of a transaction.

Use the 'retrieveDB' sample program to retrieve the stored data.

Before running this program, run loadExamplesData.[sh|cmd] to pre-populate
the DBXML container with the appropriate data. Then, run buildDB to populate
the database with the appropriate data.

When you run this program, identify the directory where you told loadExamplesData
to place the sample data:

    -h <dbenv directory> << std::endl;
EOM
    exit( -1 );
}


my $theDB = "testBerkeleyDB";
my $theContainer = "namespaceExampleData.dbxml";
while (@ARGV && $ARGV[0] =~ /^-/ )
{
    my $opt = shift @ARGV ;

    if ($opt =~ s/-h//)
      { $path2DbEnv = $opt || shift }
    else
      { usage() }
}

usage()
  if ! $path2DbEnv ;


eval {
    # open a container in the db environment
    my $env = new DbEnv(0);
    $env->set_cachesize(0, 64 * 1024, 1);
    $env->open($path2DbEnv,
                Db::DB_INIT_MPOOL|Db::DB_CREATE|Db::DB_INIT_LOCK|Db::DB_INIT_LOG|Db::DB_INIT_TXN);
    my $theMgr = new XmlManager($env);

    # Open a database in the environment
    my $myDb openedDatabase($theDB, $mgr );

    my $containerTxn = $theMgr->createTransaction();
    my $container = $theMgr->openContainer($containerTxn, $theContainer);
    $containerTxn->commit();

    try
    {
      //Get a transaction
      XmlTransaction txn = mgr.createTransaction();

            XmlQueryContext resultsContext = mgr.createQueryContext();

            //Find all the vendor documents.
	    std::string fullQuery = "collection('" + container.getName() + "')/vendor";
            XmlResults results( mgr.query(txn, fullQuery, resultsContext) );

            resultsContext.setReturnType( XmlQueryContext::DeadValues );

            //Loop through the result set, retrieving the data we are interested in for each document.
            XmlValue value;
            while( results.next(value) )
            {
                //Query the retrieved document for the salesrep's name
                XmlQueryExpression doc_expr = mgr.prepare("salesrep/name/text()", resultsContext);
                XmlResults docResult = doc_expr.execute(value, resultsContext);

                //The result set from the document query should be of size 1. If not, the
                //document schema has changed and the query must be modified..
                assert( docResult.size() == 1 ); 

                //Pull the value out of the document query result set.
                XmlValue docValue;
                docResult.next(docValue);
                std::string theSalesRepKey = docValue.asString();

                //Reconstruct the key used to store the information relevant to this document
                Dbt theKey( (void *)theSalesRepKey.c_str(), theSalesRepKey.length() + 1 );

                //This next structure is used to hold the data that is
                // retrieved from the db.
                Dbt theData;

                //Get the data and report the results
                openedDatabase.getDatabase().get(txn.getDbTxn(), &theKey, &theData, 0 );
                std::cout << "For key: '" << (char *)theKey.get_data() << "', retrieved:" << std::endl;
                std::cout << "\t" << (char *)theData.get_data() << std::endl;
            }

            txn.commit();
    }
    catch(DbException &e)
    {
        std::cerr << "Error writing to database: "<< std::endl;
        std::cerr << e.what() << std::endl;
        std::cerr << "Aborting transaction and exiting."<< std::endl;

        exit( -1 );
    }
    //Catches XmlException
    catch(std::exception &e)
    {
        std::cerr << "Error writing to database: "<< std::endl;
        std::cerr << e.what() << std::endl;
        std::cerr << "Aborting transaction and exiting."<< std::endl;

        exit( -1 );
    }

    return 0;
}
